### ISSUE TITLE
[Contributor Welcome] Implement C++ API version of torch.nn.functional.gumbel_softmax

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#27078]
## Context 

We would like to add `torch::nn::functional::gumbel_softmax` to the C++ API, so that C++ users can easily find the equivalent of Python API `torch.nn.functional.gumbel_softmax`.

## Steps

* Add `torch::nn::GumbelSoftmaxOptions` to `torch/csrc/api/include/torch/nn/options/activation.h` (add this file if it doesn’t exist), which should include the following parameters (based on https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.gumbel_softmax)
    * `TORCH_ARG(double, tau) = 1.0`
    * `TORCH_ARG(bool, hard) = false`
    * `TORCH_ARG(double, eps) = 1e-10`
    * `TORCH_ARG(int64_t, dim) = -1`
    * NOTE: please make sure to add the same comments for the parameters as in the Python version. For example, for parameter `tau` we should say `non-negative scalar temperature`
* Add `torch::nn::functional::gumbel_softmax(...)` in  `torch/csrc/api/include/torch/nn/functional/activation.h` (add this file if it doesn’t exist). The function should have the following signature:

```cpp
namespace torch {
namespace nn {
namespace functional {

inline Tensor gumbel_softmax(
    const Tensor& logits,
    const GumbelSoftmaxOptions& options = {}) {
  ...
}

} // namespace functional
} // namespace nn
} // namespace torch
```

* Add test for `torch::nn::functional::gumbel_softmax(...)` in test/cpp/api/functional.cpp. It can just check whether the function output is as expected for the following cases (expressed in Python):

```python
>>> logits = torch.randn(20, 32)
>>> # Sample soft categorical using reparametrization trick:
>>> F.gumbel_softmax(logits, tau=1, hard=False)
>>> # Sample hard categorical using "Straight-through" trick:
>>> F.gumbel_softmax(logits, tau=1, hard=True)
```

## Helpful Resources

There are quite a few PRs for adding new functionals / new modules for the C++ API (the list of PRs is in https://github.com/pytorch/pytorch/issues/25883), which can serve as great references. Also please ping @yf225 on this issue if you encounter any problems.

## How do I claim this feature request?

Please comment in this issue if you are interested in working on it.

cc @yf225

**Comment by pushkalkatara** [comment]
Hi @yf225 I am interested to work on the issue.

**Comment by yf225** [comment]
@pushkalkatara Thanks! Please let me know if I can help in any way :)

**Comment by Naresh1318** [comment]
@yf225  May I also have a crack at this?

**Comment by yf225** [comment]
@pushkalkatara Are you currently working on this issue?

**Comment by mrsalehi** [comment]
@yf225 Is anyone currently working on this issue?

**Comment by yf225** [comment]
@Naresh1318 : Thanks a lot for your interest! It seems that @pushkalkatara didn't respond, would you be interested in picking up this issue?

@mrsalehi : Thanks a lot for your interest as well! I will let @Naresh1318 take a crack at it first since he posted earlier. If you are interested, we also have many layers / functions left to be done at https://github.com/pytorch/pytorch/issues/25883, and `Flatten` might be a great one to start with :D Thanks again!

**Comment by Naresh1318** [comment]
Yes. I’d be interested in working on this!

**Comment by yf225** [comment]
@Naresh1318 Thanks! I just assigned the issue to you :D

**Comment by Naresh1318** [comment]
Hey @yf225,

Thanks for assigning the issue!

I was trying to write the last test case as seen here https://github.com/pytorch/pytorch/blob/848d1ba13a3f77bb6f14656f2d84e2228e37a8ef/test/test_nn.py#L8913 But, I wasn't able to find a function call having a parameter similar to `delta` in c++.

**Comment by yf225** [comment]
@Naresh1318 I think `torch::allclose(logits_soft.grad, logits_hard.grad, tol)` should work awesome for this use case. Thanks a lot for your help!

**Comment by Naresh1318** [comment]
Thanks for that.

I tried adding that but the test seems to fail. Here's the python test function and the c++ one that I have written 
https://github.com/pytorch/pytorch/blob/848d1ba13a3f77bb6f14656f2d84e2228e37a8ef/test/test_nn.py#L8898

```c++
    { // Test 5: Grad
        auto logits_soft = torch::zeros({10, 10});
        logits_soft.set_requires_grad(true);
        auto logits_hard = torch::zeros({10, 10});
        logits_hard.set_requires_grad(true);

        auto seed = torch::default_rng_seed_val;
        auto y_soft = F::gumbel_softmax(logits_soft, GumbelSoftmaxOptions().hard(false));
        torch::manual_seed(seed);
        auto y_hard = F::gumbel_softmax(logits_hard, GumbelSoftmaxOptions().hard(true));

        y_soft.sum().backward();
        y_hard.sum().backward();

        // 2eps = 1x addition + 1x subtraction
        std::cout << logits_soft.grad() << std::endl;
        std::cout << logits_hard.grad() << std::endl;
        ASSERT_TRUE(torch::allclose(logits_soft.grad(), logits_hard.grad(), 2.384185791015625e-07));
    }
```
Could you let me know if there's anything c++ equivalent to `torch.random.get_rng_state()`, `torch.random.set_rng_state(seed)` and `torch.finfo()`.

 Also, could you please point me to a c++ test that uses variables for different dtypes like 
https://github.com/pytorch/pytorch/blob/848d1ba13a3f77bb6f14656f2d84e2228e37a8ef/test/test_nn.py#L8915
https://github.com/pytorch/pytorch/blob/848d1ba13a3f77bb6f14656f2d84e2228e37a8ef/test/test_nn.py#L8916

**Comment by yf225** [comment]
@Naresh1318 my apologies for the late reply and thanks so much for writing the detailed tests! Currently we don't have `torch.random.get_rng_state()` / `torch.random.set_rng_state(seed)` / `torch.finfo()` in the C++ API, but it's on our list to add them. I suspect that it's probably difficult to write tests for `gumbel_softmax` that are as comprehensive as the Python tests, and I think just checking that some common configurations give the correct value (by comparing with the Python output value) would be sufficient for now. Thanks a lot again for your help!

**Comment by Naresh1318** [comment]
@yf225, Thanks for the clarification. Here's my PR: https://github.com/pytorch/pytorch/pull/28121

**Comment by chrahunt** [comment]
This was addressed in #27713.
